lib/spack/spack/stage.py

```diff
@@ -378,7 +378,9 @@ def fetch(self, mirror_only=False):
             # Join URLs of mirror roots with mirror paths. Because
             # urljoin() will strip everything past the final '/' in
             # the root, so we add a '/' if it is not present.
-            mirror_roots = [root if root.endswith('/') else root + '/'
+            mirror_roots = [canonicalize_path(root, make_abs=False)
```

Comment:
Adding `make_abs=False` just in this spot isn't particularly readable.  also, why call it twice like this?  DRY... use a variable.

Reply:
More importantly, if this is a relative path, what's it relative to, and is that documented somewhere?

This is coming from the mirror config, and the working directory for fetchers is the stage path.  But the stage could be anywhere (the stage location is user configurable), so a relative path from there is not reliable.  If it's relative to anything, it should probably be relative to the config file it came from, or to the spack installation, or something predictable that a user would know about.  It seems like to me that this will get really confusing for users.

Reply:
It's not a relative path, it's a url and when we try to turn urls into absolute paths we get garbage.

`file:///path/to/mirror` becomes `/path/to/spack/file:///path/to/mirror`.

Reply:
The refactoring you suggested should take care of the legibility issue, though.